DOC:  How-to guides for using MS Azure and AWS S3 as backends

LaureVergeron opened the pull request:
See http://s3-server.readthedocs.io/en/doc-aws_s3_and_ms_azure_backends/USING_PUBLIC_CLOUDS/# for preview

JianqinWang commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+- :code:`legacyAwsBehavior` : set to :code:`true` to indicate this region should
+  behave like AWS S3 :code:`us-east-1` region, set to :code:`false` to indicate
+  this region should behave like any other AWS S3 region;
+- :code:`bucketName` : set to an *existing bucket* in your AWS S3 Account; this
```
The user will have to make sure that the bucket both exists and has versioning enabled. I see it mentioned in line 151 but I think it makes more sense to have it mentioned here in the beginning of the guide.

LaureVergeron replied:
I added a section saying "From the AWS Console", and stating what should be done. Please let me know if it's good for you

JianqinWang commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+locationConfig.json
+~~~~~~~~~~~~~~~~~~~
+
+Add a new region, which should describe the MS Azure hosted container you'll be
```
Same comment as in the AWS S3 section: "Edit this file to add a new region..."

LaureVergeron replied:
Fixed

JianqinWang commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+writing your data to. There are a few configurable options here:
+
+- :code:`type` : set to :code:`azure` to indicate this region is hosted in MS Azure;
+- :code:`legacyAwsBehavior` : set to :code:`false`;
```
'legacyAwsBehavior' might need to be 'true', @dora-korpar ?

dora-korpar replied:
Yeah, we have all Azure backends set to 'true' for consistency during testing. It doesn't really matter either way what it's set to; it's more for our S3 to return errors the way the 'us-east-1' AWS region does. Just depends on what functionality the user wants

LaureVergeron replied:
OK, interesting. I've done all my tests with legacyAwsBehavior set to false, hence why I set it to that there. I'll just add the description I put for AWS S3 then, sounds good?

JianqinWang commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+
+   # Create a local bucket hosting data in AWS S3
+   $> s3cmd --host=127.0.0.1:8000 mb s3://zenkontainer --region=azure-test
+   # Put an object to AWS S3, and store the metadata locally
```
"MS Azure" instead of "AWS S3"?

LaureVergeron replied:
Fixed, thanks

JianqinWang commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+
+.. code:: shell
+
+   # Create a local bucket hosting data in AWS S3
```
'MS Azure' instead of 'AWS S3'? (also in line 236)

LaureVergeron replied:
Fixed, thanks

LaurenSpiegel replied:
storing data in AWS S3.  I don't like the term "hosted" because it is confusing with AWS's hosted style buckets.

LaureVergeron replied:
Changed both

dora-korpar commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+Testing: put an object to MS Azure using CloudServer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to start testing pushing to MS Azure, you will need to create a local
```
Again, same as with AWS, you don't need to create a bucket in the Azure location for the objects to be put there. It can be done at the object level

dora-korpar commented:
It might also be useful to note somewhere that if they link a pre-existing bucket/container from one of the clouds with Cloudserver, any data that was previously stored there won't be accessible through Cloudserver

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+locationConfig.json
+~~~~~~~~~~~~~~~~~~~
+
+Add a new region, which should describe the AWS S3 hosted bucket you'll be
```
Fixed

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+------------
+
+As stated in our `GETTING STARTED guide <../GETTING_STARTED/#location-configuration>`__,
+new backends can be added by creating a region with the right endpoint and
```
I've put both wordings as people are reasonably familiar with the concept of AWS regions usually, so I think it's good if it echoes to that concept for them. Do you agree?

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+- :code:`legacyAwsBehavior` : set to :code:`true` to indicate this region should
+  behave like AWS S3 :code:`us-east-1` region, set to :code:`false` to indicate
+  this region should behave like any other AWS S3 region;
+- :code:`bucketName` : set to an *existing bucket* in your AWS S3 Account; this
```
Done :)

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+.. code:: shell
+
+    [zenko]
+    aws_access_key_id=WHDBFKILOSDDVF78NPMQ
```
Yes, of course :) the SAS is a dummy too

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+  same in your local bucket and your AWS S3 bucket; set to :code:`false` if you
+  want your object name to be of the form :code:`{{localBucketName}}/{{objectname}}`
+  in your AWS S3 hosted bucket
+- :code:`credentialsProfile` : set to the profile name allowing you to access your
```
I don't know if we are, that's probably a question for @LaurenSpiegel ?
I think advocating for using profiles is more scalable, though, as that means they can put all their profiles in one .aws/credentials file and then configure their regions with the different credentials more easily (it's easier to check a human-readble name matches, than randomly generated id/signatures), don't you think?

LaurenSpiegel replied:
We are allowing both.  In federation, we will be using the flat accessKey, secretKey approach since it is easier than having people create another file in the environment.  So, we should have an example of both.

LaureVergeron replied:
Added the credentials option :)

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+   # List locally to check you have the metadata
+   $> s3cmd --host=127.0.0.1:8000 ls s3://zenkobucket
+    2017-10-23 10:26       330   s3://zenkobucket/testput
+
```
I believe this is an interesting usecase, but it's a bit more complex than "one bucket - one region". I would add a separate section in the documentation about that usecase, does that sound OK?
Also, in that situation, your local bucket would still only have the metadata for those objects that are going to a public cloud, or would the data be replicated both locally and in the chosen cloud?

dora-korpar replied:
Yup, local bucket will only have the metadata. And yeah, I'm ok with the object location constraint stuff being in a separate section, I just want to make sure that functionality is documented as well :)

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to start testing pushing to AWS S3, you will need to create a local
+bucket in the AWS S3 region - this local bucket will only store the metadata
```
Changed here

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+Troubleshooting
+~~~~~~~~~~~~~~~
+
+Make sure your :code:`~/.s3cfg` file has credentials matching your local
```
I've added a link to our template in this section so they can find it. I'll keep the host option so they understand this acts as a gateway (this question kept coming up at 42), if that's OK with you?

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,255 @@
+Microsoft Azure as a backend
+----------------------------
+
+locationConfig.json
```
Added :)

ironman-machine commented:
PR has been updated. Reviewers, please be cautious.

LaureVergeron commented:
@dora-korpar @JianqinWang : addressed all your comments, thanks for reviewing.
@dora-korpar : added warning about preexisting objects. Will add section about one bucket hosting objects in multiple backends if that sounds good?

LaureVergeron commented:
@dora-korpar : added a section about setting location at object level as suggested
@LaurenSpiegel : added the an example of how to set credentials.accessKey and credentials.secretKey for an aws_s3 type region
PR rebased on master. Thanks both and @JianqinWang for the reviews

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+
+Edit the :code:`restEndpoint` section of your :code:`config.json` file to add
+an endpoint definition matching your new AWS S3 hosted region. Following our
+previous example, it would look like:
```
Hasn't this become obsolete for public clouds? I did a docker run without mounting config.json and could write to a custom AWS location... @LaurenSpiegel ?

LaurenSpiegel replied:
You have to modify your config.json if you want to add the new locationConstraint 'aws-test' as the default for a certain endpoint. I think it is confusing to use the endpoint 's3.amazonaws.com' and this addition is unnecessary.  As long as you have some default location for each of your endpoints, you don't need to add more just because you add a new location.

LaureVergeron replied:
OK, I added an "IMPORTANT" section specifying that you shouold only change config.json to change the default location to an endpoint. In this way, we document how to do it, and it should be less confusing. Let me know.

dora-korpar commented:
Thanks for adding the object level location stuff Laure, but it looks like we're going to deprecate that feature, so we're not going to need it :# my bad!

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+Using Public Clouds as backends
```
as data backends

LaureVergeron replied:
Changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+------------
+
+As stated in our `GETTING STARTED guide <../GETTING_STARTED/#location-configuration>`__,
+new backends can be added by creating a region (also called location constraint)
```
new data

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+- `Amazon S3 <#aws-s3-as-a-backend>`__ ;
+- `Microsoft Azure <#microsoft-azure-as-a-backend>`__ .
+
+For all Public Cloud backends, you'll have to do some editing inside the
```
For each public cloud backend, you will have to edit your CloudServer locationConfig.json and do a few setup steps on the applicable public cloud backend.

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+   This is an advanced usecase, and it will require that you set your backends
+   as explained below. Once this is done, you may refer to the
+   `backend configuration at object level <#backend-configuration-at-object-level>`__
+   section of this document.
```
please remove this.  I want to deprecate this feature if possible.

LaureVergeron replied:
Great news, it's been really confusing to ppl when explained :) removed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+
+Create a bucket where you will host your data for this new location constraint.
+This bucket must have versioning enabled. This is an option you may choose to
+activate at step 2 of Bucket Creation in the console.
```
the aws console? Perhaps, just instruct them to set up an AWS bucket with their client of choice and enable versioning?

LaureVergeron replied:
I added pointers for AWS CLI as well as the Console info; from what I see at events, ppl either use the Console or AWS CLI, so it might be user friendly to give pointers for these two tools.

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+Create a bucket where you will host your data for this new location constraint.
+This bucket must have versioning enabled. This is an option you may choose to
+activate at step 2 of Bucket Creation in the console.
+In this example, our bucket will be name ``zenkobucket`` and has versioning
```
named

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+locationConfig.json
+^^^^^^^^^^^^^^^^^^^
+
+Edit this file to add a new region, which should describe the AWS S3 hosted
```
Edit this file to add a new location constraint. This location constraint will contain the information for the AWS S3 bucket to which you will be writing your data whenever you create a CloudServer bucket in this location.

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+Edit this file to add a new region, which should describe the AWS S3 hosted
+bucket you'll be writing your data to. There are a few configurable options here:
+
+- :code:`type` : set to :code:`aws_s3` to indicate this region is hosted in AWS S3;
```
to indicate this location constraint is AWS S3

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+  behave like AWS S3 :code:`us-east-1` region, set to :code:`false` to indicate
+  this region should behave like any other AWS S3 region;
+- :code:`bucketName` : set to an *existing bucket* in your AWS S3 Account; this
+  is the bucket that will host all the data written in this region;
```
this is the bucket in which your data will be stored for this location constraint

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+    (...)
+
+.. WARNING::
+   If you set :code:`bucketMatch` to :code:`true`, we strongly advise for having
```
we strongly advise that you only have one local bucket...

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+
+.. WARNING::
+   If you set :code:`bucketMatch` to :code:`true`, we strongly advise for having
+   only one local bucket per AWS S3 location. Indeed, since your objects names
```
Without bucketMatch set to false, your object names in your AWS S3 bucket will not be prefixed with your Cloud Server bucket name. This means that if you put an object `foo` to your CloudServer bucket `zenko1` and you then put a different `foo` to your CloudServer bucket `zenko2` and both `zenko1` and `zenko2` point to the same AWS bucket, the second `foo` will overwrite the first `foo`.

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+
+.. TIP::
+   If you set the :code:`credentials` object in your
+   :code:`locationConfig.json` file, you may skip this section
```
why can you skip?  because you set the accessKey/secretKey directly?  unclear.

LaureVergeron replied:
I think that, because `credentials` is an object when `credentialsProfile` is a string entry in the dict, the wording will avoid confusion (and it's documented two sections before that). I can specify `credentials.accessKey` and `credentials.secretKey` if you think that will be clearer.

LaurenSpiegel replied:
I think it would be clearer to say -- If you explicitly set your accessKey and secretKey in in the credentials section of your aws_s3 location, then you do not need a credentials file in your home directory and can skip the next step.

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Inside the repository, once all the files have been edited, you should be able
+to start the server and start testing pushing to AWS S3.
```
and start writing data to AWS S3 through CloudServer

LaureVergeron replied:
Changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. TIP::
+   If you set the :code:`credentials` object in your
```
with accessKey and secretKey?

LaureVergeron replied:
See comment above

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+   :code:`.aws/credentials` file
+
+Mount all the files that have been edited to override defaults, and do a
+standard Docker run; then you can start testing pushing to AWS S3.
```
you can start writing data to AWS S3 through your CloudServer

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+
+In order to start testing pushing to AWS S3, you will need to create a local
+bucket in the AWS S3 location constraint - this local bucket will only store the
+metadata locally, while both the data and the metadata will be stored on AWS S3.
```
the data and any user metadata (x-amz-meta headers sent with an put object) and tags

LaureVergeron replied:
changed

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+From your Stoage Account dashboard, create a container where you will host your
+data for this new location constraint.
+
+You will also need to generate a SAS (Shared Access Signature) from the Console.
```
we will now be using a storage account name and key.  pr from @electrachong coming likely tomorrow.

LaureVergeron replied:
Changed; @electrachong , can you confirm the example uses the right format, please?

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+Once you click "Generate SAS", a "SAS token" and a "Blob service SAS URL" will
+appear. You want to use the **SAS token without the initial question mark**.
+
+In this example, our container will be name ``zenkontainer``, and will belong to
```
will be called

LaureVergeron replied:
changed to "named" for consistency with your earlier comment on AWS S3 as a data backend

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,459 @@
+locationConfig.json
+^^^^^^^^^^^^^^^^^^^
+
+Edit this file to add a new region, which should describe the MS Azure hosted
```
same comments to the azure section below as aws section above.

LaureVergeron replied:
Applied all applicable

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+
+.. IMPORTANT::
+   You only need to follow this section if you want to define your new location
+   :code:`aws-test` as the default for AWS S3
```
as the default for AWS S3? as the default for an endpoint...

LaureVergeron replied:
Well, in the context of this guide, it's for s3.amazonaws.com, which I shortened to AWS S3. I think it might be easier to comprehend that with the example. I agree though we could document that config.json should only be modified to add defaults for endpoints. Maybe in a more generic / less "how-to" oriented file?

LaurenSpiegel replied:
I would say "as the default for a specific endpoint (e.g., 'zenkotos3.com')"

And then use zenkotos3.com as the example below and remove references to default AWS S3 location. I just find it really confusing to say we have a default AWS S3 location.

LaureVergeron replied:
Edited, and moved that section to the end as it is not specific to S3

electrachong commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+From your Storage Account dashboard, create a container where you will host your
+data for this new location constraint.
+
+You will also need to get one of your Storage Account Access Keys, and to
```
typo: `and to`

LaureVergeron replied:
? isn't that what is there?

electrachong replied:
Oh, I was reading it quickly and thought `and to` was a typo, but I think it actually works fine.

electrachong commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+This can be found from your Storage Account dashboard, under "Settings, then
+"Access keys".
+
+In this example, our container will be named ``zenkontainer``, and will belong
```
cute pun

LaureVergeron replied:
:)

electrachong commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+  behave like AWS S3 :code:`us-east-1` region, set to :code:`false` to indicate
+  this region should behave like any other AWS S3 region (in the case of MS Azure
+  hosted data, this is mostly relevant for the format of errors);
+- :code:`azureBlobEndpoint` : set to your storage account's endpoint, usually
```
updated to `azureStorageEndpoint`

LaureVergeron replied:
fixed

electrachong commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+  :code:`{{localBucketName}}/{{objectname}}` in your MS Azure container ;
+- :code:`azureStorageAccountName` : the MS Azure Storage Account to which your
+  container belongs;
+- :code:`azureStorageAccessKey` : one of the AccessKeys associated to the above
```
*`Access Keys`

LaureVergeron replied:
fixed

electrachong commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+access key is :code:`accessKey1` and the secret key is :code:`verySecretKey1`.
+For more informations, refer to our template `~/.s3cfg <./CLIENTS/#s3cmd>`__ .
+
+Pre existing objects in your AWS S3 hosted bucket can unfortunately not be
```
I think usually this is `Preexisting` or `Pre-existing`

LaureVergeron replied:
thanks

electrachong commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+access key is :code:`accessKey1` and the secret key is :code:`verySecretKey1`.
+For more informations, refer to our template `~/.s3cfg <./CLIENTS/#s3cmd>`__ .
+
+Pre existing objects in your MS Azure container can unfortunately not be
```
`Preexisting` or `Pre-existing`

electrachong commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+	  "azureStorageAccessKey": "auhyDo8izbuU4aZGdhxnWh0ODKFP3IWjsN1UfFaoqFbnYzPj9bxeCVAzTIcgzdgqomDKx6QS+8ov8PYCON0Nxw=="
+	}
+    },
+    (...)
```
Optionally, users may set the azureStorageAccountName, azureStorageAccessKey, and azureStorageEndpoint as environment variables in the following form:
```
$ export azure-test_AZURE_STORAGE_ACCOUNT_NAME="zenkomeetups"
$ export azure-test_AZURE_STORAGE_ACCESS_KEY="auhyDo8izbuU4aZGdhxnWh0ODKFP3IWjsN1UfFaoqFbnYzPj9bxeCVAzTIcgzdgqomDKx6QS+8ov8PYCON0Nxw=="
$ export azure-test_AZURE_STORAGE_ENDPOINT="https://zenkomeetups.blob.core.windows.net/"
```
If they specify these details in both environmental variables and the locationConfig.json, the values stored in the environmental variables take priority and are the ones that will be used.

I don't know if it's important to document this detail, it could be we created these environmental variables primarily for our own testing purposes. But since we have this option...

LaureVergeron replied:
I added a "TIP" box, I think it's relevant indeed

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+This section of the documentation shows you how to set up our currently
+supported public cloud backends:
+
+- `Amazon S3 <#aws-s3-as-a-backend>`__ ;
```
fixed, well spotted

LaureVergeron commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,374 @@
+supported public cloud backends:
+
+- `Amazon S3 <#aws-s3-as-a-backend>`__ ;
+- `Microsoft Azure <#microsoft-azure-as-a-backend>`__ .
```
fixed, well spotted

LaureVergeron commented:
Hello all,
I think I've addressed all your comments. This is rebased and ready to merge unless something else shows up :)

LaurenSpiegel commented on docs/USING_PUBLIC_CLOUDS.rst:
```diff
@@ -0,0 +1,388 @@
+^^^^^^^^^^^^^^^^^^
+
+.. TIP::
+   If you explicitly set your :code:`accessKey` and :code:`secretKey` the 
```
missing the word "in"

LaureVergeron replied:
Fixed

LaureVergeron commented:
@LaurenSpiegel : should have addressed everything, let me know :)
